### PR TITLE
[ALLUXIO-2409] Startup consistency check

### PR DESCRIPTION
**Pull request by calvinjia**
https://alluxio.atlassian.net/browse/ALLUXIO-2409

This is an initial implementation of the feature. The check runs automatically on startup and has an indicator in the web ui based on its status and results. If no files are inconsistent, the line in red does not show up.

Example screenshot:
screen shot 2016-11-07 at 10 59 49 am

There are several improvements we could make to the feature which I've listed as subtasks in the JIRA ticket.

### CODE REVIEW

**Comment by calvinjia** [comment]
@gpang @apc999

File: core/server/src/main/java/alluxio/master/file/FileSystemMaster.java [hunk 1]
```diff
@@ -248,7 +251,7 @@ public static String getJournalDirectory(String baseDirectory) {
    */
   public FileSystemMaster(BlockMaster blockMaster, Journal journal) {
     this(blockMaster, journal, ExecutorServiceFactories
-        .fixedThreadPoolExecutorServiceFactory(Constants.FILE_SYSTEM_MASTER_NAME, 2));
+        .fixedThreadPoolExecutorServiceFactory(Constants.FILE_SYSTEM_MASTER_NAME, 3));
```
Comment (apc999) [inline core/server/src/main/java/alluxio/master/file/FileSystemMaster.java]: will this cause the file system to spend one more thread after startup consistency check?

Comment (calvinjia) [inline core/server/src/main/java/alluxio/master/file/FileSystemMaster.java]: Yes, the thread will just be left in the pool. An alternative way would be to use a separate thread, but I think this thread can be reused in the future if we want to do something like update the status.

Comment (apc999) [inline core/server/src/main/java/alluxio/master/file/FileSystemMaster.java]: LGTM

**Comment by apc999** [comment]
is it necessary to have `sbin/` now to be different from `bin/` for commands that only admins should run, like `checkConsistency`?

**Comment by calvinjia** [comment]
@apc999 I think it makes sense to organize our shell tools better. There are several tools currently which require admin privileges, ie. start-all with SudoMount. In addition, the alluxio command contains many cluster management utilities such as rsyncing as well as dev tools like generating thrift or proto. I've created https://alluxio.atlassian.net/browse/ALLUXIO-2422 to track this.

File: core/server/src/main/java/alluxio/master/file/FileSystemMaster.java [hunk 2]
```diff
@@ -421,10 +424,32 @@ public void start(boolean isLeader) throws IOException {
...
+        @Override
+        public List<AlluxioURI> call() throws Exception {
+          return checkConsistency(new AlluxioURI("/"), CheckConsistencyOptions.defaults());
+        }
+      });
     }
   }
 
   /**
+   * @return The list of inconsistent files at start up, null if the check has not completed or
```
Comment (aaudiber) [inline core/server/src/main/java/alluxio/master/file/FileSystemMaster.java]: lowercase with no period for *consistency* with the rest of the file

File: core/server/src/main/java/alluxio/master/file/FileSystemMaster.java [hunk 3]
```diff
@@ -421,10 +424,32 @@ public void start(boolean isLeader) throws IOException {
...
+   *         failed.
+   */
+  public List<AlluxioURI> getStartupConsistencyCheck() {
+    boolean isDone = mStartupConsistencyCheck.isDone();
```
Comment (aaudiber) [inline core/server/src/main/java/alluxio/master/file/FileSystemMaster.java]: could this be inlined?

File: core/server/src/main/java/alluxio/master/file/FileSystemMaster.java [hunk 4]
```diff
@@ -421,10 +424,32 @@ public void start(boolean isLeader) throws IOException {
...
+    if (isDone) {
+      try {
+        return mStartupConsistencyCheck.get();
+      } catch (Exception e) {
+        // fall through
```
Comment (aaudiber) [inline core/server/src/main/java/alluxio/master/file/FileSystemMaster.java]: should we log this? when is it not a real problem?

File: tests/src/test/java/alluxio/master/StartupConsistencyCheckTest.java [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+
+import javax.annotation.Nullable;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Test the consistency check which happens on master start up.
```
Comment (aaudiber) [inline tests/src/test/java/alluxio/master/StartupConsistencyCheckTest.java]: nit: class-level comment is usually written as a description, not imperative tense

File: tests/src/test/java/alluxio/master/StartupConsistencyCheckTest.java [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    Assert.assertEquals(expected, result);
+  }
+
+  /**
+   * Waits for the startup consistency check to complete with a limit of 1 minute.
+   * @param master the file system master which is starting up
+   */
+  private void waitForStartupConsistencyCheck(final FileSystemMaster master) {
+    CommonUtils.waitFor("Startup consistency check completion", new Function<Void, Boolean>() {
+      @Nullable
```
Comment (aaudiber) [inline tests/src/test/java/alluxio/master/StartupConsistencyCheckTest.java]: remove `@Nullable`

File: tests/src/test/java/alluxio/master/StartupConsistencyCheckTest.java [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+      @Override
+      public Boolean apply(Void aVoid) {
+        return master.getStartupConsistencyCheck() != null;
+      }
+    }, 60000);
```
Comment (aaudiber) [inline tests/src/test/java/alluxio/master/StartupConsistencyCheckTest.java]: `Constants.MINUTE_MS`

File: tests/src/test/java/alluxio/master/StartupConsistencyCheckTest.java [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+    List expected = Lists.newArrayList(TOP_LEVEL_FILE, SECOND_LEVEL_DIR, THIRD_LEVEL_FILE);
+    List result = master.getStartupConsistencyCheck();
+    Collections.sort(expected);
+    Collections.sort(result);
+    Assert.assertEquals(expected, result);
+  }
+
+  /**
+   * Waits for the startup consistency check to complete with a limit of 1 minute.
+   * @param master the file system master which is starting up
```
Comment (aaudiber) [inline tests/src/test/java/alluxio/master/StartupConsistencyCheckTest.java]: nit: blank line before `@param` section

File: core/server/src/main/java/alluxio/master/file/FileSystemMaster.java [hunk 9]
```diff
@@ -421,10 +424,32 @@ public void start(boolean isLeader) throws IOException {
       mLostFilesDetectionService = getExecutorService().submit(new HeartbeatThread(
           HeartbeatContext.MASTER_LOST_FILES_DETECTION, new LostFilesDetectionHeartbeatExecutor(),
           Configuration.getInt(PropertyKey.MASTER_HEARTBEAT_INTERVAL_MS)));
+      mStartupConsistencyCheck = getExecutorService().submit(new Callable<List<AlluxioURI>>() {
```
Comment (gpang) [inline core/server/src/main/java/alluxio/master/file/FileSystemMaster.java]: When will this check start running? Before or after the master recovers from the journal?

Is the master serving during the check? This means only read only operations would be possible during the check?

Also, how does this interact with standby masters becoming leader masters?

File: core/server/src/main/java/alluxio/master/file/FileSystemMaster.java (hunk 2 above)
Comment (calvinjia) [inline core/server/src/main/java/alluxio/master/file/FileSystemMaster.java]: Updated, thanks.

File: core/server/src/main/java/alluxio/master/file/FileSystemMaster.java (hunk 3 above)
Comment (calvinjia) [inline core/server/src/main/java/alluxio/master/file/FileSystemMaster.java]: Updated, thanks.

File: core/server/src/main/java/alluxio/master/file/FileSystemMaster.java (hunk 4 above)
Comment (calvinjia) [inline core/server/src/main/java/alluxio/master/file/FileSystemMaster.java]: Good point, I've added a log message. I've also made a follow up task to expose this info to the user.

File: tests/src/test/java/alluxio/master/StartupConsistencyCheckTest.java (hunk 8 above)
Comment (calvinjia) [inline tests/src/test/java/alluxio/master/StartupConsistencyCheckTest.java]: Updated, thanks.

File: tests/src/test/java/alluxio/master/StartupConsistencyCheckTest.java (hunk 6 above)
Comment (calvinjia) [inline tests/src/test/java/alluxio/master/StartupConsistencyCheckTest.java]: Updated, thanks.

File: tests/src/test/java/alluxio/master/StartupConsistencyCheckTest.java (hunk 7 above)
Comment (calvinjia) [inline tests/src/test/java/alluxio/master/StartupConsistencyCheckTest.java]: Updated, thanks.

File: core/server/src/main/java/alluxio/master/file/FileSystemMaster.java (hunk 9 above)
Comment (calvinjia) [inline core/server/src/main/java/alluxio/master/file/FileSystemMaster.java]: This check starts running only after the master is fully operational, so after journal recovery. The master is serving during the check, but only read operations would be permitted (ie. safe mode). When a standby master assumes leadership, the check will be rerun.

Comment (gpang) [inline core/server/src/main/java/alluxio/master/file/FileSystemMaster.java]: ahhh, I see. I thought it was in the constructor but I was mistaken. This all makes sense now.

**Pull request merged**